File: scala/57.scala
import scala.collection.mutable.ListBuffer
object Solution {
    def insert(intervals: Array[Array[Int]], newInterval: Array[Int]): Array[Array[Int]] = {
        val i = intervals.lastIndexWhere(p => p.head < newInterval.head)       
        val arr = if (i == -1) {
            newInterval +: intervals
        } else {
            Array(intervals(i), newInterval) ++ intervals.slice(i+1, intervals.length)
        }
        def overlap(arr1: Array[Int], arr2: Array[Int]): Boolean = {
            return arr1.last >= arr2.head
        } 
        val res = ListBuffer[Array[Int]]()
        (arr :+ Array(100001, 100001)).foldLeft(arr.head) {
            case (last, cur) => {
                if (overlap(last, cur)) {
                    last(1) = math.max(last.last, cur.last)
                    last
                } else {
                    res += last
                    cur
                }
            }
        }
        intervals.slice(0, i) ++ res.toArray
    }
}

File: leetcodeJava/57.java
package leetcodeJava;

import java.util.*;
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        var intervalList = new ArrayList<int[]>(Arrays.asList(intervals));
        int i;
        for (i = 0; i < intervals.length; i++) {
            if (intervals[i][0] >= newInterval[0]) {
                break;
            }
        }
        intervalList.add(i, newInterval);
        int start = intervalList.get(i)[0], end = intervalList.get(i)[1], j = i;
        if (i > 0 && newInterval[0] <= intervalList.get(i-1)[1]) {
            start = Math.min(start, intervalList.get(i-1)[0]);
            i--;
        }
        for (j = i; j < intervalList.size(); j++) {
            if (intervalList.get(j)[0] <= newInterval[1]) {
                end = Math.max(end, intervalList.get(j)[1]);
            } else {
                break;
            }
        }
        var flag = false;
        var res = new LinkedList<int[]>();
        for (var k = 0; k < intervalList.size(); k++) {
            if (k < i || k >= j) {
                res.add(intervalList.get(k));
            } else if (!flag) {
                res.add(new int[]{start, end});
                flag = true;
            }
        }
        return res.toArray(new int[0][]);
    }
}